android/src/main/java/com/swmansion/gesturehandler/core/FlingGestureHandler.kt

```diff
@@ -35,26 +34,74 @@ class FlingGestureHandler : GestureHandler<FlingGestureHandler>() {
     handler!!.postDelayed(failDelayed, maxDurationMs)
   }
 
-  private fun tryEndFling(event: MotionEvent) = if (
-    maxNumberOfPointersSimultaneously == numberOfPointersRequired &&
-    (
-      direction and DIRECTION_RIGHT != 0 &&
-        event.rawX - startX > minAcceptableDelta ||
-        direction and DIRECTION_LEFT != 0 &&
-        startX - event.rawX > minAcceptableDelta ||
-        direction and DIRECTION_UP != 0 &&
-        startY - event.rawY > minAcceptableDelta ||
-        direction and DIRECTION_DOWN != 0 &&
-        event.rawY - startY > minAcceptableDelta
+  private fun tryEndFling(event: MotionEvent): Boolean {
+    data class SimpleVector(val x: Double, val y: Double)
+
+    fun toSafeNumber(unsafe: Double): Double = if (unsafe.isFinite()) unsafe else 0.0
+
+    fun toUnitVector(vec: SimpleVector): SimpleVector {
+      val magnitude = hypot(vec.x, vec.y)
+      // toSafeNumber protects against division by zero
+      return SimpleVector(
+        toSafeNumber(vec.x / magnitude),
+        toSafeNumber(vec.y / magnitude)
       )
-  ) {
-    handler!!.removeCallbacksAndMessages(null)
-    activate()
-    true
-  } else {
-    false
-  }
+    }
+
+    fun compareSimilarity(
+      vecA: SimpleVector,
+      vecB: SimpleVector
+    ): Double {
+      // inputs have to be unit vectors
+      // returns scalar on range from -1.0 to 1.0
+      return vecA.x * vecB.x + vecA.y * vecB.y
+    }
+
+    fun compareAlignment(
+      vec: SimpleVector,
+      directionVec: SimpleVector,
+      direction: Int
+    ): Boolean =
+      compareSimilarity(vec, directionVec) > minDirectionAlignment &&
+        (this.direction and direction != 0)
+
+    val velocityTracker = VelocityTracker.obtain()
+    addVelocityMovement(velocityTracker, event)
+    velocityTracker!!.computeCurrentVelocity(1000)
+
+    val velocityVector = SimpleVector(
+      velocityTracker.xVelocity.toDouble(),
+      velocityTracker.yVelocity.toDouble()
+    )
+
+    val velocityUnitVector = toUnitVector(velocityVector)
 
+    val alignmentList = arrayOf(
+      compareAlignment(velocityUnitVector, SimpleVector(-1.0, 0.0), DIRECTION_LEFT),
+      compareAlignment(velocityUnitVector, SimpleVector(1.0, 0.0), DIRECTION_RIGHT),
+      compareAlignment(velocityUnitVector, SimpleVector(0.0, -1.0), DIRECTION_UP),
+      compareAlignment(velocityUnitVector, SimpleVector(0.0, 1.0), DIRECTION_DOWN)
+    )
```

Comment:
Maybe I'm dumb, but I don't see checking if the alignment is consistent with what the user passed in `direction` prop. It seems like we're always checking all four directions and activating on any of them.

Reply:
Hello, thank you for the whole review :)
Consistence with user input is checked inside `compareAlignment()`.
We use this array because the user may request more than 1 directions to be 'activatory', as `direction` is a bitmask. 
As you suggested, I will separate this code into a separate class, which will resolve most of the issues present here.